xds/src/main/java/io/grpc/xds/XdsClient.java

```diff
@@ -0,0 +1,320 @@
+import com.google.common.base.Stopwatch;
+import com.google.common.base.Supplier;
+import com.google.protobuf.InvalidProtocolBufferException;
+import io.envoyproxy.envoy.api.v2.ClusterLoadAssignment;
+import io.envoyproxy.envoy.api.v2.DiscoveryRequest;
+import io.envoyproxy.envoy.api.v2.DiscoveryResponse;
+import io.envoyproxy.envoy.api.v2.core.Node;
+import io.envoyproxy.envoy.service.discovery.v2.AggregatedDiscoveryServiceGrpc;
+import io.grpc.ManagedChannel;
+import io.grpc.Status;
+import io.grpc.SynchronizationContext;
+import io.grpc.SynchronizationContext.ScheduledHandle;
+import io.grpc.internal.BackoffPolicy;
+import io.grpc.stub.StreamObserver;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.CheckForNull;
+
+final class XdsClient {
+  private static Logger logger = Logger.getLogger(XdsClient.class.getName());
+
+  private final ManagedChannel channel;
+  private final SynchronizationContext syncCtx;
+  private final ScheduledExecutorService timerService;
+  private final BackoffPolicy.Provider backoffPolicyProvider;
+  private final Supplier<Stopwatch> stopwatchSupplier;
+  private final Node xdsNode;
+  private final String authority;
+
+  private final Map<String, EndpointWatcher> endpointWatchers = new LinkedHashMap<>();
+
+  // never null
+  private AdsStream adsStream;
+  // never null
+  private BackoffPolicy adsRpcRetryPolicy;
+  @CheckForNull
+  private ScheduledHandle adsRpcRetryTimer;
+
+  XdsClient(
+      ManagedChannel channel,
+      SynchronizationContext syncCtx,
+      ScheduledExecutorService timerService,
+      BackoffPolicy.Provider backoffPolicyProvider,
+      Supplier<Stopwatch> stopwatchSupplier,
+      Node xdsNode,
+      final String authority) {
+    this.channel = checkNotNull(channel, "channel");
+    this.syncCtx = checkNotNull(syncCtx, "syncCtx");
+    this.timerService = checkNotNull(timerService, "timerService");
+    this.backoffPolicyProvider = checkNotNull(backoffPolicyProvider, "backoffPolicyProvider");
+    this.adsRpcRetryPolicy = backoffPolicyProvider.get();
+    this.stopwatchSupplier = checkNotNull(stopwatchSupplier, "stopwatchSupplier");
+    this.xdsNode = checkNotNull(xdsNode, "xdsNode");
+    this.authority = checkNotNull(authority, "authority");
+
+    syncCtx.executeLater(
+        new Runnable() {
+          @Override
+          public void run() {
+            logger.log(Level.INFO, "Starting XdsClient {0}", XdsClient.this);
+            adsStream = new AdsStream();
+          }
+        }
+    );
+  }
+
+  void shutdown() {
+    channel.shutdown();
+    adsStream.shutdown();
+    cancelRetryTimer();
+
+  }
+
+  void watchEndpointData(String cluster, EndpointWatcher endpointWatcher) {
+    checkArgument(
+        !endpointWatchers.containsKey(cluster),
```

Comment:
I thought we are allowing multiple watchers for the same cluster based on this comment https://github.com/grpc/grpc-java/pull/6303#discussion_r337065675

Reply:
Changed. Thanks for pointing it out. I'm not sure then how the graceful swap will be implemented yet though.